Source/Core/VideoCommon/VertexLoaderARM64.cpp

```diff
@@ -476,33 +475,61 @@ void VertexLoaderARM64::GenerateVertexLoader()
   {
     static const u8 map[8] = {7, 6, 15, 14};
     const u8 scaling_exponent = map[u32(m_VtxAttr.g0.NormalFormat.Value())];
-    const int limit = m_VtxAttr.g0.NormalElements == NormalComponentCount::NTB ? 3 : 1;
 
-    std::optional<s32> offset = std::nullopt;
-    for (int i = 0; i < limit; i++)
+    const int elem_size = GetElementSize(m_VtxAttr.g0.NormalFormat);
+
+    const int load_bytes = elem_size * 3;
+    const int load_size = GetLoadSize(load_bytes);
+
+    // Normal
+    std::optional<s32> offset = GetAddressImm(CPArray::Normal, m_VtxDesc.low.Normal,
+                                              EncodeRegTo64(scratch1_reg), load_size << 3);
+    ReadVertex(m_VtxDesc.low.Normal, m_VtxAttr.g0.NormalFormat, 3, 3, true, scaling_exponent,
+               &m_native_vtx_decl.normals[0], offset);
+
+    if (m_VtxAttr.g0.NormalElements == NormalComponentCount::NTB)
     {
-      if (!i || m_VtxAttr.g0.NormalIndex3)
+      const bool index3 = IsIndexed(m_VtxDesc.low.Normal) && m_VtxAttr.g0.NormalIndex3;
+
+      // Tangent
+      // If in Index3 mode, and indexed components are used, replace the index with a new index.
+      if (index3)
       {
-        int elem_size = GetElementSize(m_VtxAttr.g0.NormalFormat);
+        offset = GetAddressImm(CPArray::Normal, m_VtxDesc.low.Normal, EncodeRegTo64(scratch1_reg),
+                               load_size << 3);
+      }
+      // The tangent comes after the normal; even in index3 mode, this offset is applied.
+      // Note that this is different from adding 1 to the index, as the stride for indices may be
+      // different from the size of the tangent itself.
+      if (!offset.has_value())
+        ADD(EncodeRegTo64(scratch1_reg), EncodeRegTo64(scratch1_reg), load_bytes);
+      else
+        *offset += load_bytes;
 
-        int load_bytes = elem_size * 3;
-        int load_size = GetLoadSize(load_bytes);
+      ReadVertex(m_VtxDesc.low.Normal, m_VtxAttr.g0.NormalFormat, 3, 3, true, scaling_exponent,
+                 &m_native_vtx_decl.normals[1], offset);
 
+      // Undo the offset above so that data points to the normal instead of the tangent.
+      // This way, we can add 2*elem_size below to always point to the binormal, even if we replace
+      // data with a new index (which would point to the normal).
+      if (!offset.has_value())
+        SUB(EncodeRegTo64(scratch1_reg), EncodeRegTo64(scratch1_reg), load_bytes);
+      else
+        *offset -= load_bytes;
```

Comment:
Could you make it so ReadVertex takes a parameter for which register to use? That way we could make the ADD earlier write the result to scratch3_reg so that scratch1_reg doesn't get overwritten, saving us from having to do a SUB here.

Also, isn't this SUB redundant if index3 is true?

Reply:
Perhaps best to restructure `offset` into some kind of type that can have either a register or an integer... `std::variant`, maybe?

Reply:
I did an initial implementation of this [here](https://github.com/pokechu22/dolphin/commits/archive/vertex-loader-direct-normals-with-index3-variant-1), but on further thought, I think VertexLoaderARM64 is more complicated than it needs to be. I'm currently writing some unit tests so that I can make sure the behavior makes sense, and then I'll revise further.

Reply:
Vertices are always small enough that the source and destination offsets (from the start of the current vertex) are less than 255, so `LDUR`/`SDUR` can always be used. I've refactored is so that `ReadVertex` always takes both the register and an offset from it, and `GetVertexAddr` returns both a register and an offset (for indexed vertex components, this is `EncodeRegTo64(scratch1_reg)` (a pointer to the right offset in the array corresponding to that vertex component) and an offset of 0, while for direct components this is `src_reg` (the start of the current vertex) and an offset of `m_src_ofs` (which is the offset from the start of the vertex to the component we're reading). For the tangent and binormal I just add the extra offset to that offset, and everything works cleanly. There's no need to generate `ADD` or `SUB` instructions to change the offset at all anymore.